### PR TITLE
Add more tslint rules

### PR DESCRIPTION
**Pull request by fabioberger**
## Description

This PR adds the following TSLint rules:

**custom-no-magic-numbers**: No magic numbers to improve code readability. Instead, assign the magic number to a clearly named variable. The one exception is when instantiating a BigNumber instance (e.g `const amount = new BigNumber(5);` is valid)

**no-switch-case-fall-through**: One must add a bread for each base in a switch statement. This is to avoid accidentally forgetting to add the break.

**boolean-naming**: Boolean variables names must be prefixed with one of the following: 'is', 'does', 'should', 'was', 'has', 'can', 'did', 'would'. This is to improve readability, especially during code reviews.

## Motivation and Context

We have quite a [few coding conventions](https://quip.com/RtvjA8EIUKBB) that aren't yet enforced by the linter. This PR adds a few of them to the linter.

## How Has This Been Tested?

Ran the linter against all packages.

## Types of changes

*   [ ] Bug fix (non-breaking change which fixes an issue)
*   [ ] New feature (non-breaking change which adds functionality)
*   [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

*   [ ] Change requires a change to the documentation.
*   [ ] Added tests to cover my changes.
*   [ ] Added new entries to the relevant CHANGELOG.jsons.
*   [ ] Labeled this PR with the 'WIP' label if it is a work in progress.
*   [ ] Labeled this PR with the labels corresponding to the changed package.

### CODE REVIEW

File: packages/testnet-faucets/src/ts/constants.ts [hunk 1]
```diff
@@ -0,0 +1,5 @@
+export const constants = {
```
Comment (LogvinovLeon) [inline packages/testnet-faucets/src/ts/constants.ts]: Let's use that https://www.npmjs.com/package/http-status

File: packages/tslint-config/package.json [hunk 2]
```diff
@@ -44,7 +44,8 @@
         "lodash": "^4.17.4",
         "tslint": "5.8.0",
         "tslint-eslint-rules": "^4.1.1",
-        "tslint-react": "^3.2.0"
+        "tslint-react": "^3.2.0",
+        "tsutils": "^2.12.1"
```
Comment (LogvinovLeon) [inline packages/tslint-config/package.json]: Where do we use that?

File: packages/utils/src/abi_decoder.ts [hunk 3]
```diff
@@ -23,7 +23,8 @@ export class AbiDecoder {
             formatted = formatted.slice(2);
         }
 
-        formatted = _.padStart(formatted, 40, '0');
+        const desiredLength = 40;
```
Comment (LogvinovLeon) [inline packages/utils/src/abi_decoder.ts]: address length

**Comment by albrow** [comment]
Hmm... while some of the "magic number" changes make a lot of sense, I'm not sure that lines like `const sixthIndex = 6;` and `const num = 4;` actually make the code more readable. IMO whether something is a magic number that could benefit from a name is more of a case-by-case thing better suited for addressing in code review. Is this really something we want to enforce indiscriminately?

As an alternative, I would be down for enforcing a linter rule for any non-zero numbers which appear in the code more than 2-3 times. But I'm not sure if that's possible.

File: packages/contract-wrappers/test/order_validation_test.ts [hunk 4]
```diff
@@ -422,7 +426,8 @@ describe('OrderValidation', () => {
                 takerAddress,
                 zrxTokenAddress,
             );
-            expect(transferFromAsync.callCount).to.be.equal(4);
+            const expectedCallCount = 4;
+            expect(transferFromAsync.callCount).to.be.equal(expectedCallCount);
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/contract-wrappers/test/order_validation_test.ts]: I don't think so... but let's just disable it for the ones like this where it doesn't make sense.

File: packages/monorepo-scripts/src/postpublish_utils.ts [hunk 5]
```diff
@@ -158,6 +158,7 @@ export const postpublishUtils = {
             // HACK: tsconfig.json needs wildcard directory endings as `/**/*`
             // but TypeDoc needs it as `/**` in order to pick up files at the root
             if (_.endsWith(includePath, '/**/*')) {
+                // tslint:disable-next-line:custom-no-magic-numbers
                 includePath = includePath.slice(0, -2);
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/monorepo-scripts/src/postpublish_utils.ts]: Eh. I think this is fine.

File: packages/order-utils/test/signature_utils_test.ts [hunk 6]
```diff
@@ -110,10 +112,12 @@ describe('Signature utils', () => {
                     if (payload.method === 'eth_sign') {
                         const [address, message] = payload.params;
                         const signature = await web3Wrapper.signMessageAsync(address, message);
+                        // tslint:disable-next-line:custom-no-magic-numbers
+                        const rsvHex = `0x${signature.substr(130)}${signature.substr(2, 128)}`;
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/order-utils/test/signature_utils_test.ts]: Agreed, but I don't actually know what it means 😬

File: packages/subproviders/src/utils/wallet_utils.ts [hunk 7]
```diff
@@ -30,10 +30,10 @@ class DerivedHDKeyInfoIterator implements IterableIterator<DerivedHDKeyInfo> {
             baseDerivationPath,
             derivationPath: fullDerivationPath,
         };
-        const done = this._index === this._searchLimit;
+        const isDone = this._index === this._searchLimit;
         this._index++;
         return {
-            done,
+            done: isDone,
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply packages/subproviders/src/utils/wallet_utils.ts]: It is enforced by the `IterableIterator` interface.

File: packages/tslint-config/package.json (hunk 2 above)
Comment (fabioberger) [reply packages/tslint-config/package.json]: In `customNoMagicNumbersRule.ts`

**Comment by fabioberger** [comment]
@albrow I agree that setting magic numbers to variables does not always improve readability. That is why we did not enable this rule previously. That being said, I've been having to comment a lot on PR with vague uses of magic numbers. 

Let's try enabling the rule so as to warn the developer while coding that they should double-check that their use of a magic number is still readable. I am find with heavy use of:

```
// tslint:disable-next-line:custom-no-magic-numbers
```

in places where introducing a variable would hurt readability. Let's not just introduce variables to satisfy the rule.

Is that a fair compromise?

File: packages/contract-wrappers/test/ether_token_wrapper_test.ts [hunk 8]
```diff
@@ -154,6 +153,7 @@ describe('EtherTokenWrapper', () => {
             const preWETHBalance = await contractWrappers.token.getBalanceAsync(wethContractAddress, addressWithETH);
             expect(preWETHBalance).to.be.bignumber.equal(0);
 
+            // tslint:disable-next-line:custom-no-magic-numbers
             const overWETHBalance = preWETHBalance.add(999999999);
```
[the remark this answers is not in the document]
Comment (LogvinovLeon) [reply packages/contract-wrappers/test/ether_token_wrapper_test.ts]: We should change the number to 1. The only important thing here is that it should be positive.

**Comment by albrow** [comment]
I usually strive to have a set of linter rules that do not require frequent exceptions. On the other hand, if this helps reduce code review time based on your personal experience, it sounds like it's worth it 😄

**Pull request merged**